### DISCUSSION TITLE
Wrapping a `response_model` Pydantic class in `@dataclass` breaks OpenAPI doc generation when you use the response model the 2nd time

### DISCUSSION THREAD

**Discussion by JLHasson** [fastapi/fastapi#8667] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional
import uvicorn
import pydantic
from fastapi import FastAPI
from dataclasses import dataclass

app = FastAPI()


@dataclass
class ResponseModel(pydantic.BaseModel):
    hello: str


@app.get("/", response_model=ResponseModel)
def read_root():
    return {"hello": "World"}


@app.get("/items/{item_id}", response_model=ResponseModel)
def read_item(item_id: int, q: Optional[str] = None):
    return {"item_id": item_id, "q": q}


if __name__ == "__main__":
    uvicorn.run(
        "test_fastapi:app",
        host="0.0.0.0",
        port=8080,
        reload=True,
    )
```

### Description

Run `python test_fastapi.py` and navigate to http://localhost:8080/docs#/. You should see the following:
<img width="636" alt="image" src="https://user-images.githubusercontent.com/2087380/139576171-517e62a0-4a0b-4f09-9a47-0fc3efa27f25.png">

This bug only appears when you have _more than one_ route returning `ResponseModel`.

When you wrap the Pydantic model in `@dataclass` it causes `get_flat_models_from_routes` 
https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/openapi/utils.py#L326
to return two different classes for `ResponseModel`, when in fact there should only be one. `model_name_map`
https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/openapi/utils.py#L386
will only contain one of these two classes as a key, causing the following KeyError to occur:
```
  File "/home/.../venv/lib/python3.9/site-packages/fastapi/applications.py", line 136, in openapi
    self.openapi_schema = get_openapi(
  File "/home/.../venv/lib/python3.9/site-packages/fastapi/openapi/utils.py", line 387, in get_openapi
    definitions = get_model_definitions(
  File "/home/.../venv/lib/python3.9/site-packages/fastapi/utils.py", line 34, in get_model_definitions
    model_name = model_name_map[model]
KeyError: <class 'pydantic.dataclasses.ResponseModel'>
```

I believe this is a FastAPI error and not a Pydantic error since `flat_models` contains the incorrect values, but it's possible that `model_name_map` should contain the second class in which case this would be a Pydantic bug.

Removing the `@dataclass` fixes the error. While maybe it should be obvious that you should not wrap Pydantic models in `@dataclass` the system should likely still be tolerant of this behavior (or error with a more meaningful message).

### Operating System

Linux

### FastAPI Version

0.70.0

### Python Version

3.9.5

**Comment by clichedmoog** [comment]
I am using 0.71.0 and It works but has an error when showing the swagger doc.

**Comment by rmtmckenzie** [comment]
For anyone else running across this, changing from `from dataclasses import dataclass` to `from pydantic.dataclasses import dataclass` fixed the issur for me.

**Comment by YuriiMotov** [comment, marked as answer]
Works fine in current version (0.115.13)
